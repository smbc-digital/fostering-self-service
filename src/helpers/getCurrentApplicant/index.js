import { Applicant, SECOND_APPLICANT_ROUTE } from 'constants'

const getCurrentApplicant = match => {
    if (match.params && match.params[0] === SECOND_APPLICANT_ROUTE) {
        return Applicant.SecondApplicant
    }

    return Applicant.FirstApplicant
}

export default getCurrentApplicant