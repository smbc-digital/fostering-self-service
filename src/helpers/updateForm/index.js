import { fetchWithTimeout } from '../index'

const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

const StageName = {
    HomeVisit: 0,
    Application: 1
}

const HomeVisitFormName = {
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguagesSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 8
}

const ApplicationFormName = {
    References: 0,
    GpDetails: 1,
    AddressHistory: 2
}

const getHomeVisitUpdateEndpoint = form => {
    switch (form) {
        case HomeVisitFormName.TellUsAboutYourself:
            return '/fostering/home-visit/about-yourself'
        case HomeVisitFormName.YourEmploymentDetails:
            return '/fostering/home-visit/your-employment-details'
        case HomeVisitFormName.LanguagesSpokenInYourHome:
            return '/fostering/home-visit/languages-spoken-in-your-home'
        case HomeVisitFormName.YourFosteringHistory:
            return '/fostering/home-visit/your-fostering-history'
        case HomeVisitFormName.YourPartnership:
            return '/fostering/home-visit/partnership-status'
        case HomeVisitFormName.TellUsAboutYourInterestInFostering:
            return '/fostering/home-visit/interest-in-fostering'
        case HomeVisitFormName.YourHealth:
            return '/fostering/home-visit/about-your-health'
        case HomeVisitFormName.YourHousehold:
            return '/fostering/home-visit/household'
        case HomeVisitFormName.ChildrenLivingAwayFromYourHome:
            return '/fostering/home-visit/children-living-away-from-home'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getApplicationUpdateEndpoint = form => {
    switch (form) {
        case ApplicationFormName.References:
            return '/fostering/application/references'
        case ApplicationFormName.GpDetails:
            return '/fostering/application/gp-details'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getFormStageEndpoint = stage => {
    switch (stage) {
        case StageName.HomeVisit: 
            return '/fostering/home-visit'
        case StageName.Application:
            return '/fostering/application'
        default:
            throw new Error('No matching endpoint for given stage.')
    }
}

const reduceProperties = object => Object.keys(object).reduce((acc, property) => {
    return {
        ...acc,
        [property]: object[property].value
    }
}, {})

const updateForm = async (endpoint, formData) => {
    const parsedFormData = parseFormData(formData)

    const response = await fetchWithTimeout(endpoint, {
        method: 'PATCH',
        credentials: 'include',
        body: JSON.stringify(parsedFormData),
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }
    }, 30000)

    if (!response.ok) {
        throw Error(response.statusText)
    }

    return await response.json()
}

const updateFormStatus = ({ form, stage, currentStatus, setStatus }) => {
    if (currentStatus === TaskStatus.None) {
        fetchWithTimeout(`${getFormStageEndpoint(stage)}/status`,
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                }),
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                }
            }, 30000)

        setStatus(TaskStatus.NotCompleted)
    }
}

const parseFormData = ({ firstApplicant, secondApplicant, familyReference, firstPersonalReference, secondPersonalReference, ...formData }) => {
    let parsedObject = reduceProperties(formData)

    if (firstApplicant) {
        parsedObject.firstApplicant = reduceProperties(firstApplicant)
    }

    if (secondApplicant) {
        parsedObject.secondApplicant = reduceProperties(secondApplicant)
    }

    if (familyReference) {
        parsedObject.familyReference = reduceProperties(familyReference)
        parsedObject.familyReference.address = familyReference.address.value
    }

    if (firstPersonalReference) {
        parsedObject.firstPersonalReference = reduceProperties(firstPersonalReference)
        parsedObject.firstPersonalReference.address = firstPersonalReference.address.value
    }

    if (secondPersonalReference) {
        parsedObject.secondPersonalReference = reduceProperties(secondPersonalReference)
        parsedObject.secondPersonalReference.address = secondPersonalReference.address.value
    }

    return parsedObject
}

const updateApplicationForm = async (form, formData) => {
    const endpoint = getApplicationUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

const updateHomeVisitForm = async (form, formData) => {
    const endpoint = getHomeVisitUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

export {
    TaskStatus,
    StageName,
    HomeVisitFormName,
    ApplicationFormName,
    updateFormStatus,
    parseFormData,
    updateApplicationForm,
    updateHomeVisitForm
}