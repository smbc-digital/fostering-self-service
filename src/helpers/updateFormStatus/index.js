export const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

export const FormName =
{
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguageSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 7
}

export default function (form, currentStatus, setStatus) {
    if (currentStatus === TaskStatus.None) {
        fetch('/front-end-endpoint-update-status',
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                })
            })

        setStatus(TaskStatus.NotCompleted)
    }
}