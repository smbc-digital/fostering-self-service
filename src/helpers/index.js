import fetchWithTimeout from './fetchWithTimeout'
import { getPageRoute } from './pagehelper'
import getCurrentApplicant from './getCurrentApplicant'
import { updateFormStatus, updateForm, FormName, TaskStatus } from './updateForm'

export {
    fetchWithTimeout,
    getPageRoute,
    getCurrentApplicant,
    updateFormStatus,
    updateForm,
    FormName,
    TaskStatus
}