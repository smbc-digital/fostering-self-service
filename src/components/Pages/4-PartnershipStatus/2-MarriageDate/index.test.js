import { React, mount, useContextMock, renderer } from '../../../../helpers/SetupTest'
import MarriageDate from './index'
import * as helpers from '../../../../helpers'

describe('MarriageDate', () => {

    const onChangeStatusMock = jest.fn()

    beforeEach(() => {
        useContextMock.mockReturnValue({
            onChange: jest.fn(),
            onChangeStatus: onChangeStatusMock,
            dateOfMarriage: '26/06/2019'
        })
    }) 

    it('should call updateForm', async () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        
        helpers.updateForm = jest.fn().mockReturnValue(Promise.resolve())
        const wrapper = mount(<MarriageDate history={history} />)
        
        // Act
        await wrapper.find('form').simulate('submit')
        await Promise.resolve()

        // Assert
        expect(helpers.updateForm).toHaveBeenCalled()
    })

    it('should push to next page on submit', async () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        
        helpers.updateForm = jest.fn().mockReturnValue(Promise.resolve(0))
        const wrapper = mount(<MarriageDate history={history} />)
        
        // Act
        await wrapper.find('form').simulate('submit')
        await Promise.resolve()

        // Assert
        expect(history.push).toHaveBeenCalledWith(helpers.getPageRoute(11))
    })

    it('should push to start page', async () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        
        helpers.updateForm = jest.fn().mockReturnValue(Promise.resolve(0))
        const wrapper = mount(<MarriageDate history={history} />)
        
        // Act
        await wrapper.find('Button').at(1).simulate('click')
        await Promise.resolve()

        // Assert
        expect(history.push).toHaveBeenCalledWith(helpers.getPageRoute(1))
    })

    it('should call updateForm on save and go back', async () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        
        helpers.updateForm = jest.fn().mockReturnValue(Promise.resolve(0))
        const wrapper = mount(<MarriageDate history={history} />)
        
        // Act
        await wrapper.find('Button').at(1).simulate('click')
        await Promise.resolve()

        // Assert
        expect(helpers.updateForm).toHaveBeenCalled()
    })

    it('should call onChangeStatus', async () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        
        helpers.updateForm = jest.fn().mockReturnValue(Promise.resolve(0))
        const wrapper = mount(<MarriageDate history={history} />)
        
        // Act
        await wrapper.find('form').simulate('submit')
        await Promise.resolve()

        // Assert
        expect(onChangeStatusMock).toHaveBeenCalled()
    })

    it('should push to error page on updateForm error', async () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        
        helpers.updateForm = jest.fn().mockImplementation(() => { throw new Error() })
        const wrapper = mount(<MarriageDate history={history} />)
        
        // Act
        wrapper.find('form').simulate('submit')

        // Assert
        expect(history.push).toHaveBeenCalledWith('/error')
    })

    it('should set isLoading', async () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        
        helpers.updateForm = jest.fn().mockReturnValue(Promise.resolve(0))
        const wrapper = mount(<MarriageDate history={history} />)
        
        // Act
        await wrapper.find('form').simulate('submit')
        await Promise.resolve()

        // Assert
        expect(wrapper.find('Button').at(0).props().isLoading).toBe(true)
    })

    describe('snapshot', () => {

        it('should render correctly', () => {
            // Act
            const tree = renderer
                .create(<MarriageDate history={{}} />)
                .toJSON()

            // Assert
            expect(tree).toMatchSnapshot()
        })
    })
})