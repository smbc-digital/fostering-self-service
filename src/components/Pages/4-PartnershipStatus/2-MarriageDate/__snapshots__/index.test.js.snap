// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarriageDate snapshot should render correctly 1`] = `
Array [
  <h1>
    Your fostering journey
  </h1>,
  <h2>
    Your partnership status
  </h2>,
  <form
    onSubmit={[Function]}
  >
    <h3
      className="optional no-margin-bottom"
    >
      Date of your marriage or civil partnerhsip
    </h3>
    <fieldset
      className="input-memorable-date"
    >
      <p>
        For example, 31 3 1980
      </p>
      <div>
        <label>
          Day
        </label>
        <input
          aria-label="day"
          className=""
          name="day"
          onChange={[Function]}
          onKeyDown={[Function]}
          type="number"
          value=""
        />
      </div>
      <div>
        <label>
          Month
        </label>
        <input
          aria-label="month"
          className=""
          name="month"
          onChange={[Function]}
          onKeyDown={[Function]}
          type="number"
          value=""
        />
      </div>
      <div>
        <label>
          Year
        </label>
        <input
          aria-label="year"
          className=""
          name="year"
          onChange={[Function]}
          onKeyDown={[Function]}
          type="number"
          value=""
        />
      </div>
    </fieldset>
    <button
      className="button-primary"
      disabled={false}
      id={undefined}
      onClick={undefined}
      type="submit"
    >
      Save and next step
    </button>
    <p>
      <a
        href="#"
        onClick={[Function]}
      >
        Back
      </a>
    </p>
    <button
      className="button-inverted left-chevron"
      disabled={false}
      id={undefined}
      onClick={[Function]}
      type="submit"
    >
      Save and go back to fostering area
    </button>
  </form>,
]
`;
